Common/src/linear_algebra/CSysSolve.cpp

```diff
@@ -658,6 +661,7 @@ unsigned long CSysSolve<ScalarType>::Smoother_LinSolver(const CSysVector<ScalarT
    product (A_x), for the latter two this is done only on the first call to the method. ---*/
 
   if (!smooth_ready) {
+    SU2_OMP_BARRIER
     SU2_OMP_MASTER
     {
       auto nVar = b.GetNVar();
```

Comment:
Bug fix relating to #830, sometimes the code would hang on the first iteration because of this:
```c++
if (!smooth_ready) { // if working variables have not been allocated we need to allocate...
  SU2_OMP_BARRIER // <- This is the fix.
  SU2_OMP_MASTER // ...only the master thread can do allocation as those variables are shared
  {
    // allocate variables
    ...
    smooth_ready = true; // set this flag so we don't allocate on the next call.
  }
  SU2_OMP_BARRIER // now we need a barrier so all threads "see" the new state of the variables.
  // but because we are "smart" the barrier is inside the "if" so that we don't hit it every time.
  // But if some threads arrive (very) late to the party they might see smooth_ready == true already
  // and they miss the barrier which results in a deadlock (all threads must arrive before execution
  // can continue). One solution is to make sure all threads are inside the "if" (via another barrier)
  // before changing the condition deciding whether that statement is executed.
}
```

Reply:
Oh! I think I did see this happen a couple of times when I briefly tested the multi-threading. I built SU2 with `-Dwith-omp=true`

It happened when I didn't specify a number of threads with the `-t` option. But it ran okay if I specified even `-t 1`. Does this fix that? Or do you have to specify a thread count if you build with OMP?

Reply:
When you don't use option -t (--threads) SU2 picks up the number of threads from the environment, which often is the maximum. If you use MPI at the same time like this the system will probably be oversubscribed, I noticed that OpenMPI 3.1.4 on Ubuntu 16 deadlocks when this happens (oversubscribing with just threads does not seem to deadlock).
This fix was for a more subtle problem.

Another aspect of using MPI+threads is that you need to pay attention to the binding mode, some mpi will "--bind-to core" when the number of ranks is small, this then locks all threads spawn by each rank to the core. Always use "--bind-to socket/numa" as appropriate.